Add docs for custom actions (#1248)
RoadieHQ/marketing-site: content/docs/scaffolder/self-hosted-scaffolder-actions/index.md

publishedDate: '2024-03-25T12:00:00.0Z'
description: A step-by-step tutorial on developing and utilizing custom Scaffolder Actions
---

# Custom Scaffolder Actions

Roadie provides the possibility to run self-hosted, custom scaffolder actions as part of your scaffolder template. These actions can be written using the Roadie Agent Javascript/Typescript library. You can find the library and it's configuration options from its [NPM repository](https://www.npmjs.com/package/@roadiehq/roadie-agent). This documentation shows you how to configure your self-hosted actions to Roadie and write templates that use those actions.

## Getting started

### Installing the agent library

#### Option 1 - Using a scaffolder template

Roadie has developed a scaffolder template that allows you to bootstrap a node.js monorepository, which in turn allows you to start developing, building and deploying custom scaffolder actions. You can find the Scaffolder template to get started from here: https://github.com/RoadieHQ/software-templates/blob/main/scaffolder-templates/roadie-agent-scaffolder-action/template.yaml. 

#### Option 2 - Installing the library manually

> You can also install the library manually to your project. The Roadie Agent library can be included in any node.js application and incorporated to existing projects if wanted. The library allows you to run multiple scaffolder action and custom entity providers from within your own infrastructure. To get started, you need to add the node.js library `@roadiehq/roadie-agent` as a dependency into your project and implement your wanted actions or provider using the instructions in the repository. 

Reviewer: ```suggestion
The Roadie Agent package can be included in any node.js application and incorporated to existing projects if preferred. The library allows you to run multiple scaffolder action and custom entity providers from within your own infrastructure. To get started, you need to add the node.js library `@roadiehq/roadie-agent` as a dependency into your project and implement actions or entity providers using the instructions in the repository. 
```

### Writing your first action

> The agent library is written in Typescript and thus custom actions can be written in either Javascript or using Javascript as a bridge to trigger other commands on the host system. 

Reviewer: ```suggestion
The agent library is written in Typescript and thus custom actions can be written in either Javascript or using Typescript as a bridge to trigger other commands on the host system. 
```

An example custom action, using the repository bootstrapped by the Scaffolder template mentioned above, would look something like below.

First an index file configuring the agent and some actions for it:

```javascript
// src/index.js
import { RoadieAgent, createRoadieAgentScaffolderAction } from '@roadiehq/roadie-agent';
import { addAddtionalFuncToProjectActionHandler } from './addAddtionalFuncToProjectActionHandler.js';
import { configureMonitoringActionHandler } from './configureMonitoringActionHandler.js';

RoadieAgent.fromConfig({
  server: 'https://my-tenant.broker.roadie.so',
  accept: `${process.cwd()}/config/accept.json`,
  identifier: process.env.BROKER_TOKEN // given that you have exported and env variable called BROKER_TOKEN
})
  .addScaffolderAction(createRoadieAgentScaffolderAction({

>     name: 'add-additional-functionality-to-project-action',

Reviewer: Make the name of this "hello-world". I think its clearer.

>     name: 'add-additional-functionality-to-project-action',

Reviewer: ```suggestion
    name: 'hello-world',
```

Author: Added as an extra since screenshots are mapping to the other name

>     handler: addAddtionalFuncToProjectActionHandler

Reviewer: ```suggestion
    handler: helloWorldActionHandler
```

  }))  
  .addScaffolderAction(createRoadieAgentScaffolderAction({
    name: 'configure-monitoring-action',
    handler: configureMonitoringActionHandler
  }))
  .start();
```

An example of an individual action:

```javascript
// src/addAddtionalFuncToProjectActionHandler.js
import { writeFileSync } from "fs";

export const addAddtionalFuncToProjectActionHandler = async (context) => {
  try {

>     const greeting = `Hello, ${JSON.parse(context.payload.body)['name'] || 'world!'}`

Reviewer: ```suggestion
    const greeting = `Hello, ${context.payload.body.name || 'world!'}`
```

    await context.log(greeting);

>     writeFileSync(`${context.workspacePath}/greeting.txt`, greeting);

Reviewer: Can a step return output?

Author: No, we are currently not mapping responses/outputs back to the action

| shareWorkspace | Share Workspace | This input is optional and specifies whether to share the workspace or not. Allows the custom action to modify the workspace created by other steps in the template | boolean |
| payload        | Payload         | This input is optional. An object consisting of any additional parameters that need to be passed into the custom action.                                            | object  |

```yaml
apiVersion: scaffolder.backstage.io/v1beta3
kind: Template
metadata:
  name: custom-action-with-workspace
  title: Custom Scaffolder Action with Shared Workspace
  description: Custom Scaffolder Action using a shared workspace.

spec:
  type: service
  owner: engineering
  steps:

    - id: fetchTemplate
      name: Fetch file
      action: fetch:template
      input:
        url: ./skeleton

    - id: invoke-custom-action
      name: invoke-custom-action
      action: roadie:agent
      input:
        action: add-additional-functionality-to-project-action
        shareWorkspace: true
        payload:
          first: value1

>           second: value2

Reviewer: ```suggestion
          name: Jussi
```
